File: internal/code/code.go
package code

import (
	"bytes"
	"encoding/binary"
	"fmt"
)

type (
	Instructions []byte
	Opcode       byte
)

type Definition struct {
	Name          string
	OperandWidths []int
}

//go:generate stringer -type=Opcode
const (
	OpConstant Opcode = iota
	OpTrue
	OpFalse
	OpNull
	OpList
	OpMap
	OpClosure
	OpCurrentClosure

	OpAdd
	OpSub
	OpMul
	OpDiv
	OpMod

	OpBwAnd
	OpBwOr
	OpBwXor
	OpBwNot
	OpBwLShift
	OpBwRShift

	OpAnd
	OpOr
	OpEqual
	OpNotEqual
	OpGreaterThan
	OpGreaterThanEqual
	OpIn

	OpMinus
	OpBang
	OpIndex

	OpCall
	OpConcurrentCall
	OpReturn
	OpReturnValue

	OpJump
	OpJumpNotTruthy

	OpDot
	OpDefine
	OpGetGlobal
	OpSetGlobal
	OpGetLocal
	OpSetLocal
	OpGetBuiltin
	OpGetFree
	OpLoadModule
	OpInterpolate

	OpPop
)

var definitions = map[Opcode]*Definition{
	OpConstant:       {"OpConstant", []int{2}},
	OpTrue:           {"OpTrue", []int{}},
	OpFalse:          {"OpFalse", []int{}},
	OpNull:           {"OpNull", []int{}},
	OpList:           {"OpList", []int{2}},
	OpMap:            {"OpMap", []int{2}},
	OpClosure:        {"OpClosure", []int{2, 1}},
	OpCurrentClosure: {"OpCurrentClosure", []int{}},

	OpAdd: {"OpAdd", []int{}},
	OpSub: {"OpSub", []int{}},
	OpMul: {"OpMul", []int{}},
	OpDiv: {"OpDiv", []int{}},
	OpMod: {"OpMod", []int{}},

	OpBwAnd:    {"OpBwAnd", []int{}},
	OpBwOr:     {"OpBwOr", []int{}},
	OpBwXor:    {"OpBwXor", []int{}},
	OpBwNot:    {"OpBwNot", []int{}},
	OpBwLShift: {"OpBwLShift", []int{}},
	OpBwRShift: {"OpBwRshift", []int{}},

	OpAnd:              {"OpAnd", []int{}},
	OpOr:               {"OpOr", []int{}},
	OpEqual:            {"OpEqual", []int{}},
	OpNotEqual:         {"OpNotEqual", []int{}},
	OpGreaterThan:      {"OpGreaterThan", []int{}},
	OpGreaterThanEqual: {"OpGreaterThanEqual", []int{}},
	OpIn:               {"OpIn", []int{}},

	OpMinus: {"OpMinus", []int{}},
	OpBang:  {"OpBang", []int{}},
	OpIndex: {"OpIndex", []int{}},

	OpCall:           {"OpCall", []int{1}},
	OpConcurrentCall: {"OpConcurrentCall", []int{1}},
	OpReturn:         {"OpReturn", []int{}},
	OpReturnValue:    {"OpReturnValue", []int{}},

	OpJump:          {"OpJump", []int{2}},
	OpJumpNotTruthy: {"OpJumpNotTruthy", []int{2}},

	OpDot:         {"OpDot", []int{}},
	OpDefine:      {"OpDefine", []int{}},
	OpGetGlobal:   {"OpGetGlobal", []int{2}},
	OpSetGlobal:   {"OpSetGlobal", []int{2}},
	OpGetLocal:    {"OpGetLocal", []int{1}},
	OpSetLocal:    {"OpSetLocal", []int{1}},
	OpGetBuiltin:  {"OpGetBuiltin", []int{1}},
	OpGetFree:     {"OpGetFree", []int{1}},
	OpLoadModule:  {"OpLoadModule", []int{}},
	OpInterpolate: {"OpInterpolate", []int{2, 2}},

	OpPop: {"OpPop", []int{}},
}

func (ins Instructions) String() string {
	var out bytes.Buffer

	for i := 0; i < len(ins); {
		def, err := Lookup(ins[i])
		if err != nil {
			fmt.Fprintf(&out, "Error: %s\n", err)
			continue
		}

		operands, read := ReadOperands(def, ins[i+1:])
		fmt.Fprintf(&out, "%04d %s\n", i, ins.fmtInstruction(def, operands))
		i += read + 1
	}

	return out.String()
}

func (ins Instructions) fmtInstruction(def *Definition, operands []int) string {
	var opCount = len(def.OperandWidths)

	if len(operands) != opCount {
		return fmt.Sprintf("error: operand len %d does not match defined %d\n", len(operands), opCount)
	}

	switch opCount {
	case 0:
		return def.Name

	case 1:
		return fmt.Sprintf("%s %d", def.Name, operands[0])

	case 2:
		return fmt.Sprintf("%s %d %d", def.Name, operands[0], operands[1])

	default:
		return fmt.Sprintf("error: unhandled operand count for %s\n", def.Name)
	}
}

func Lookup(op byte) (*Definition, error) {
	if d, ok := definitions[Opcode(op)]; ok {
		return d, nil
	}
	return nil, fmt.Errorf("opcode %d undefined", op)
}

// Returns the resulting bytecode from parsing the instruction.
func Make(op Opcode, operands ...int) []byte {
	def, ok := definitions[op]
	if !ok {
		return []byte{}
	}

	instructionsLen := 1
	for _, w := range def.OperandWidths {
		instructionsLen += w
	}

	instructions := make([]byte, instructionsLen)
	instructions[0] = byte(op)

	offset := 1
	for i, o := range operands {
		width := def.OperandWidths[i]

		switch width {
		case 1:
			instructions[offset] = byte(o)

		case 2:
			binary.BigEndian.PutUint16(instructions[offset:], uint16(o))

		case 4:
			binary.BigEndian.PutUint32(instructions[offset:], uint32(o))

		case 8:
			binary.BigEndian.PutUint64(instructions[offset:], uint64(o))
		}
		offset += width
	}

	return instructions
}

// Decodes the operands of a bytecode instruction.
func ReadOperands(def *Definition, ins Instructions) ([]int, int) {
	operands := make([]int, len(def.OperandWidths))
	offset := 0
	for i, width := range def.OperandWidths {
		switch width {
		case 1:
			operands[i] = int(ReadUint8(ins[offset:]))

		case 2:
			operands[i] = int(ReadUint16(ins[offset:]))

		case 4:
			operands[i] = int(ReadUint32(ins[offset:]))
		}
		offset += width
	}
	return operands, offset
}

func ReadUint8(ins Instructions) uint8 {
	return uint8(ins[0])
}

func ReadUint16(ins Instructions) uint16 {
	return binary.BigEndian.Uint16(ins)
}

func ReadUint32(ins Instructions) uint32 {
	return binary.BigEndian.Uint32(ins)
}

func ReadUint64(ins Instructions) uint64 {
	return binary.BigEndian.Uint64(ins)
}

File: internal/vm/cvm/code.c
#include <stdlib.h>
#include "code.h"

int lookup_def(enum opcode opcode, struct definition *def) {
	if (opcode >= NUM_OPCODES) {
		return 0;
	}

	*def = definitions[opcode];
	return 1;
}

static inline void put_uint16(uint8_t *code, uint16_t i) {
	code[0] = (i & (0xff << 8)) >> 8;
	code[1] = i & 0xff;
}

static inline void put_uint32(uint8_t *code, uint32_t i) {
	code[0] = (i & (0xff << 24)) >> 24;
	code[1] = (i & (0xff << 16)) >> 16;
	code[2] = (i & (0xff << 8)) >> 8;
	code[3] = i & 0xff;
}

#include <stdio.h>
size_t vmake_bcode(uint8_t **code, size_t code_len, enum opcode op, va_list operands) {
	struct definition def;

	if (!lookup_def(op, &def)) {
		puts("dio cane");
		return code_len;
	}

	int ins_len = 1;

	for (int i = 0; i < def.noperands; i++) {
		ins_len += def.opwidths[i];
	}

	int offset = code_len;
	code_len += sizeof(uint8_t) * ins_len;
	*code = realloc(*code, code_len);
	uint8_t *bcode = *code;
	bcode[offset++] = op;

	for (int i = 0; i < def.noperands; i++) {
		int width = def.opwidths[i];

		switch (width) {
			case 1: {
				uint8_t operand = va_arg(operands, uint32_t);
				bcode[offset] = operand;
				break;
			}

			case 2: {
				uint16_t operand = va_arg(operands, uint32_t);
				put_uint16(&bcode[offset], operand);
				break;
			}

			case 4: {
				uint32_t operand = va_arg(operands, uint32_t);
				put_uint32(&bcode[offset], operand);
				break;
			}
		}

		offset += width;
	}

	return code_len;
}

// Returns the resulting bytecode from parsing the instruction.
size_t make_bcode(uint8_t **code, size_t code_len, enum opcode op, ...) {
	va_list args;
	va_start(args, op);
	int pos = vmake_bcode(code, code_len, op, args);
	va_end(args);

	return pos;
}

// Decodes the operands of a bytecode instruction.
int read_operands(struct definition def, uint8_t *ins, int **operands) {
	*operands = realloc(*operands, sizeof(int) * def.noperands);
	int offset = 0;

	for (int i = 0; i < def.noperands; i++) {
		int width = def.opwidths[i];

		switch (width) {
		case 1:
			*operands[i] = read_uint8(&ins[offset]);
		case 2:
			*operands[i] = read_uint16(&ins[offset]);
		case 3:
			*operands[i] = read_uint32(&ins[offset]);
		}

		offset += width;
	}

	return offset;
}

char *opcode_str(enum opcode op) {
	static char *strings[] = {
		"op_constant",
		"op_true",
		"op_false",
		"op_null",
		"op_list",
		"op_map",
		"op_closure",
		"op_current_closure",

		"op_add",
		"op_sub",
		"op_mul",
		"op_div",
		"op_mod",

		"op_bw_and",
		"op_bw_or",
		"op_bw_xor",
		"op_bw_not",
		"op_bw_lshift",
		"op_bw_rshift",

		"op_and",
		"op_or",
		"op_equal",
		"op_not_equal",
		"op_greater_than",
		"op_greater_than_equal",

		"op_minus",
		"op_bang",
		"op_index",

		"op_call",
		"op_concurrent_call",
		"op_return",
		"op_return_value",

		"op_jump",
		"op_jump_not_truthy",

		"op_dot",
		"op_define",
		"op_get_global",
		"op_set_global",
		"op_get_local",
		"op_set_local",
		"op_get_builtin",
		"op_get_free",
		"op_load_module",
		"op_interpolate",

		"op_pop",
		"op_halt"
	};

	return strings[op];
}

struct definition definitions[NUM_OPCODES] = {
	{"op_constant", (int[1]) {2}, 1},
	{"op_true", (int[1]) {0}, 0},
	{"op_false", (int[1]) {0}, 0},
	{"op_null", (int[1]) {0}, 0},
	{"op_list", (int[1]) {2}, 1},
	{"op_map", (int[1]) {2}, 1},
	{"op_closure", (int[2]) {2, 1}, 2},
	{"op_current_closure", (int[1]) {0}, 0},

	{"op_add", (int[1]) {0}, 0},
	{"op_sub", (int[1]) {0}, 0},
	{"op_mul", (int[1]) {0}, 0},
	{"op_div", (int[1]) {0}, 0},
	{"op_mod", (int[1]) {0}, 0},

	{"op_bw_and", (int[1]) {0}, 0},
	{"op_bw_or", (int[1]) {0}, 0},
	{"op_bw_xor", (int[1]) {0}, 0},
	{"op_bw_not", (int[1]) {0}, 0},
	{"op_bw_lshift", (int[1]) {0}, 0},
	{"op_bw_rshift", (int[1]) {0}, 0},

	{"op_and", (int[1]) {0}, 0},
	{"op_or", (int[1]) {0}, 0},
	{"op_equal", (int[1]) {0}, 0},
	{"op_not_equal", (int[1]) {0}, 0},
	{"op_greater_than", (int[1]) {0}, 0},
	{"op_greater_than_equal", (int[1]) {0}, 0},

	{"op_minus", (int[1]) {0}, 0},
	{"op_bang", (int[1]) {0}, 0},
	{"op_index", (int[1]) {0}, 0},

	{"op_call", (int[1]) {1}, 1},
	{"op_concurrent_call", (int[1]) {1}, 1},
	{"op_return", (int[1]) {0}, 0},
	{"op_return_value", (int[1]) {0}, 0},

	{"op_jump", (int[1]) {2}, 1},
	{"op_jump_not_truthy", (int[1]) {2}, 1},

	{"op_dot", (int[1]) {0}, 0},
	{"op_define", (int[1]) {0}, 0},
	{"op_get_global", (int[1]) {2}, 1},
	{"op_set_global", (int[1]) {2}, 1},
	{"op_get_local", (int[1]) {1}, 1},
	{"op_set_local", (int[1]) {1}, 1},
	{"op_get_builtin", (int[1]) {1}, 1},
	{"op_get_free", (int[1]) {1}, 1},
	{"op_load_module", (int[1]) {0}, 0},
	{"op_interpolate", (int[2]) {2, 2}, 2},

	{"op_pop", (int[1]) {0}, 0},
	{"op_halt", (int[1]) {0}, 0}
};
